### ISSUE TITLE
Use uriparser and minizip as deps

### ISSUE THREAD

**Issue by ocefpaf** [conda-forge/libkml-feedstock#2]
Once we create packages for `uriparser` and `minizip` we can use them here as a dependency and reduce the extra build time.

**Comment by zklaus** [comment]
@ocefpaf, and pinging @gillins, it's been a while, do we have any progress here? I am asking because in the cmake build for gdal (conda-forge/gdal-feedstock#629) libkml doesn't find what I think are minizip symbols and I think the approach to that problem might change depending on this issue.

**Comment by ocefpaf** [comment]
Nope. I conpletely forgot about this issue. We would need to package those fist to avoid the bundled versions.

**Comment by zklaus** [comment]
Alright, I'll try to patch the cmake stuff then.

**Comment by ocefpaf** [comment]
Thanks and sorry for the extra work. I kind of abandoned the idea in this issue and I don't have the bandwidth to get back to it.

**Comment by zklaus** [comment]
No worries. Just wanted to avoid possible double work.

**Comment by akrherz** [comment]
as denoted here conda-forge/freexl-feedstock#29 , what's the development package source of `minizip`?

**Comment by h-vetinari** [comment]
@akrherz, thanks for packaging minizip!

I tried this in #24 ([CI run](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=794764&view=results)), however, it's unfortunately incompatible with the expectations of libklm - that is, libkml will _find_ our minizip, but then fail because it expects a different header structure:
```
[  1%] Building C object src/kml/base/CMakeFiles/kmlbase.dir/contrib/minizip/iomem_simple.c.o
In file included from $SRC_DIR/src/kml/base/contrib/minizip/iomem_simple.c:31:
$SRC_DIR/src/kml/base/contrib/minizip/unzip.h:67:10: fatal error: minizip/ioapi.h: No such file or directory
   67 | #include <minizip/ioapi.h> //KMR. were using local version
      |          ^~~~~~~~~~~~~~~~~
compilation terminated.
```

I checked what libkml uses if it _doesn't_ find libminizip, and it [points](https://github.com/libkml/libkml/blob/1.3.0/cmake/External_minizip.cmake#L3) to a very old version on [sourceforge](https://sourceforge.net/projects/libkml-files/files/1.3.0/). With some luck I managed to find the [sources](https://github.com/rashadkm/minizip) also on Github - they're forked from the right upstream, but very different:

![image](https://github.com/conda-forge/libkml-feedstock/assets/33685575/87a489fd-61a7-4d18-a2b6-7698afcfcc61)

IOW, it shares about 190 commits of history out of ~2150 for minizip 4.0; upon closer inspection, I believe [minizip 1.2](https://github.com/zlib-ng/minizip-ng/tree/1.2) would be compatible. We could consider building that on a branch of the minizip feedstock? That would likely mean static builds though, as we'd otherwise run into the problem of not being able to have two different shared library builds of minizip in the same environment.

**Comment by akrherz** [comment]
Oh my @h-vetinari , thanks for this research!  I notice that ioapi.h exists in [zlib contrib](https://github.com/madler/zlib/tree/master/contrib/minizip), do you think it is compatible too?  I will look more into this today

**Comment by h-vetinari** [comment]
> I notice that ioapi.h exists in [zlib contrib](https://github.com/madler/zlib/tree/master/contrib/minizip), do you think it is compatible too?

It looks like a mirror world a bit, comparing the files and the changelogs, there's some overlap, but a vast symmetric difference. Also, the zlib/contrib still gets updated in sync with the main zlib versioning, whereas zlibg-ng/minizip-ng is on a completely different version scheme now, and the libkml sources are frozen for years. Short of trying, I have no idea if using that would work TBH, but I think it'd be more promising to repackage the version I referenced.

**Comment by ocefpaf** [comment]
Makes me wonder if we should still ship gdal with libkml.

**Comment by h-vetinari** [comment]
Now getting clobber warnings and segfaults...
```
ClobberWarning: This transaction has incompatible packages due to a shared path.
  packages: conda-forge/linux-64::libkml-1.3.0-h01aab08_1016, conda-forge/linux-64::minizip-4.0.1-h0ab5242_4
  path: 'include/minizip/unzip.h'


ClobberWarning: This transaction has incompatible packages due to a shared path.
  packages: conda-forge/linux-64::libkml-1.3.0-h01aab08_1016, conda-forge/linux-64::minizip-4.0.1-h0ab5242_4
  path: 'include/minizip/zip.h'


ClobberWarning: This transaction has incompatible packages due to a shared path.
  packages: conda-forge/linux-64::libkml-1.3.0-h01aab08_1016, conda-forge/linux-64::minizip-4.0.1-h0ab5242_4
  path: 'lib/libminizip.so'
```
See https://github.com/conda-forge/pycoalescence-feedstock/pull/54

**Comment by h-vetinari** [comment]
Some more research... In addition to downloading minizip sources from [sourceforge](http://sourceforge.net/projects/libkml-files/files/1.3.0/minizip.tar.gz/download), libkml has a partially redundant copy with some additional changes under [`/src/kml/base/contrib/minizip`](https://github.com/libkml/libkml/blob/1.3.0/src/kml/base/contrib/minizip).

First, just using the builds from https://github.com/conda-forge/minizip-feedstock/pull/9 runs into some type errors.

<details>

```
[  0%] Building CXX object src/kml/base/CMakeFiles/kmlbase.dir/attributes.cc.o
[  1%] Building C object src/kml/base/CMakeFiles/kmlbase.dir/contrib/minizip/iomem_simple.c.o
In file included from $SRC_DIR/src/kml/base/contrib/minizip/iomem_simple.c:32:
$SRC_DIR/src/kml/base/contrib/minizip/iomem_simple.h:39:28: error: conflicting types for 'read_file_func'; have 'uLong (*)(void *, void *, void *, uLong)' {aka 'long unsigned int (*)(void *, void *, void *, long unsigned int)'}
   39 | typedef uLong  (ZCALLBACK *read_file_func) OF((voidpf opaque, voidpf stream, void* buf, uLong size));
      |                            ^~~~~~~~~~~~~~
In file included from $SRC_DIR/src/kml/base/contrib/minizip/unzip.h:67,
                 from $SRC_DIR/src/kml/base/contrib/minizip/iomem_simple.c:31:
$PREFIX/include/minizip/ioapi.h:77:30: note: previous declaration of 'read_file_func' with type 'read_file_func' {aka 'unsigned int (*)(void *, void *, void *, unsigned int)'}
   77 | typedef uint32_t (ZCALLBACK *read_file_func)     (voidpf opaque, voidpf stream, void* buf, uint32_t size);
      |                              ^~~~~~~~~~~~~~
$SRC_DIR/src/kml/base/contrib/minizip/iomem_simple.h:40:28: error: conflicting types for 'write_file_func'; have 'uLong (*)(void *, void *, const void *, uLong)' {aka 'long unsigned int (*)(void *, void *, const void *, long unsigned int)'}
   40 | typedef uLong  (ZCALLBACK *write_file_func) OF((voidpf opaque, voidpf stream, const void* buf, uLong size));
      |                            ^~~~~~~~~~~~~~~
$PREFIX/include/minizip/ioapi.h:78:30: note: previous declaration of 'write_file_func' with type 'write_file_func' {aka 'unsigned int (*)(void *, void *, const void *, unsigned int)'}
   78 | typedef uint32_t (ZCALLBACK *write_file_func)    (voidpf opaque, voidpf stream, const void *buf, uint32_t size);
      |                              ^~~~~~~~~~~~~~~
$SRC_DIR/src/kml/base/contrib/minizip/iomem_simple.h:42:28: error: conflicting types for 'seek_file_func'; have 'long int (*)(void *, void *, uLong,  int)' {aka 'long int (*)(void *, void *, long unsigned int,  int)'}
   42 | typedef long   (ZCALLBACK *seek_file_func) OF((voidpf opaque, voidpf stream, uLong offset, int origin));
      |                            ^~~~~~~~~~~~~~
$PREFIX/include/minizip/ioapi.h:83:30: note: previous declaration of 'seek_file_func' with type 'seek_file_func' {aka 'long int (*)(void *, void *, unsigned int,  int)'}
   83 | typedef long     (ZCALLBACK *seek_file_func)     (voidpf opaque, voidpf stream, uint32_t offset, int origin);
      |                              ^~~~~~~~~~~~~~
$SRC_DIR/src/kml/base/contrib/minizip/iomem_simple.c: In function 'mem_simple_create_file':
$SRC_DIR/src/kml/base/contrib/minizip/iomem_simple.c:226:22: warning: assignment to 'read_file_func' {aka 'unsigned int (*)(void *, void *, void *, unsigned int)'} from incompatible pointer type 'uLong (*)(void *, void *, void *, uLong)' {aka 'long unsigned int (*)(void *, void *, void *, long unsigned int)'} [-Wincompatible-pointer-types]
  226 |     api->zread_file  = mem_read;
      |                      ^
$SRC_DIR/src/kml/base/contrib/minizip/iomem_simple.c:227:22: warning: assignment to 'write_file_func' {aka 'unsigned int (*)(void *, void *, const void *, unsigned int)'} from incompatible pointer type 'uLong (*)(void *, void *, const void *, uLong)' {aka 'long unsigned int (*)(void *, void *, const void *, long unsigned int)'} [-Wincompatible-pointer-types]
  227 |     api->zwrite_file = mem_write;
      |                      ^
$SRC_DIR/src/kml/base/contrib/minizip/iomem_simple.c:229:22: warning: assignment to 'seek_file_func' {aka 'long int (*)(void *, void *, unsigned int,  int)'} from incompatible pointer type 'long int (*)(void *, void *, uLong,  int)' {aka 'long int (*)(void *, void *, long unsigned int,  int)'} [-Wincompatible-pointer-types]
  229 |     api->zseek_file  = mem_seek;
      |                      ^
make[2]: *** [src/kml/base/CMakeFiles/kmlbase.dir/build.make:90: src/kml/base/CMakeFiles/kmlbase.dir/contrib/minizip/iomem_simple.c.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [CMakeFiles/Makefile2:218: src/kml/base/CMakeFiles/kmlbase.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```

</details>

My first instinct was to rip out that `contrib` layer and just forcefully rely on our own minizip, however I noticed some issues. While it would be possible to change the only place where kml picks up its contrib folder (rather than `#include <minizip/...>` proper), there's some stuff in there that doesn't get covered by our minizip headers.

The immediate standout is this [function](https://github.com/libkml/libkml/blob/1.3.0/src/kml/base/contrib/minizip/iomem_simple.h#L73) which gets [used](https://github.com/libkml/libkml/blob/1.3.0/src/kml/base/zip_file.cc#L94) twice in the main zip_file implementation, so we need to have it (or replace it with something equivalent).

As it turns out, the state of libkml's [`ioapi.h`](https://github.com/libkml/libkml/blob/1.3.0/src/kml/base/contrib/minizip/ioapi.h) corresponds essentially one-to-one to minizip 1.01e, which isn't present in the minizip-ng repo, but can be found under `/contrib/minizip` in [zlib 1.2.2.3](https://github.com/madler/zlib/blob/v1.2.2.3/contrib/minizip/ioapi.h). The only difference is
```diff
 #define ZCLOSE(filefunc,filestream) ((*((filefunc).zclose_file))((filefunc).opaque,filestream))
 #define ZERROR(filefunc,filestream) ((*((filefunc).zerror_file))((filefunc).opaque,filestream))

+/*
+ * This declaration is from the proposed iomem_simple package found at
+ * http://code.trak.dk. See iomem_simple.c
+ */
+extern void* ZEXPORT mem_simple_create_file(zlib_filefunc_def* pzlib_filefunc_def, void* buffer, size_t buflen);

 #ifdef __cplusplus
 }
```

The page mentioned in the comment is long gone, but I found an [archive](https://web.archive.org/web/20180814170556/http://code.trak.dk/) from 2018. The associated [`zip.zip`](https://web.archive.org/web/20180814170556/http://code.trak.dk/download/zip.zip) contains zlib 1.2.1 (according to the changelog) under `/zlib`, has the minizip sources [matching](https://github.com/madler/zlib/tree/v1.2.1/contrib/minizip) minizip 1.00 under `/unzip`, and indeed, the sought-after `iomem_simple.{s,h}` and some more things under `/src.`

Looking at the comments in the original `iomem_simple.h` (since deleted in libkml), it says it is intended for accessing memory rather than files.
```
/* ioapi_mem2.c -- IO base function header for compress/uncompress .zip
   files using zlib + zip or unzip API

   This version of ioapi is designed to access memory rather than files.
   We do use a region of memory to put data in to and take it out of. We do
   not have auto-extending buffers and do not inform anyone else that the
   data has been written. It is really intended for accessing a zip archive
   embedded in an application such that I can write an installer with no
   external files. Creation of archives has not been attempted, although
   parts of the framework are present.
```
Based on this I realized that minizip actually _does_ have such a memory [API](https://github.com/zlib-ng/minizip-ng/blob/1.2/ioapi_mem.h), and I'm attempting a patch to see if we can make this work.

Further notes from that might be potential causes of issues:
* There are some other fixes that are libkml-specific (e.g. [this](https://github.com/libkml/libkml/blame/1.3.0/src/kml/base/contrib/minizip/iomem_simple.c#L141-L144) one which precedes even the original import into libkml, or some C99 fixes), but at least, those _should_ be covered by libkml tests.
* I also noticed that the structure being used in that function is different between [libmkl](https://github.com/libkml/libkml/blob/1.3.0/src/kml/base/contrib/minizip/ioapi.h#L46-L56) and [minizip 1.2](https://github.com/zlib-ng/minizip-ng/blob/1.2/ioapi.h#L85-L97)
```diff
 typedef struct zlib_filefunc_def_s
 {
     open_file_func      zopen_file;
+    opendisk_file_func  zopendisk_file;
     read_file_func      zread_file;
     write_file_func     zwrite_file;
     tell_file_func      ztell_file;
     seek_file_func      zseek_file;
     close_file_func     zclose_file;
     testerror_file_func zerror_file;
     voidpf              opaque;
 } zlib_filefunc_def;
```
However, since all the previous members are there, this should be compatible with libkml if recompiled consistently.

**Comment by zklaus** [comment]
> Makes me wonder if we should still ship gdal with libkml.

Just wanted to second @ocefpaf's comment. This all seems like an awful lot of work for a library that has apparently not received any TLC upstream since 2017.

I do immensely appreciate @h-vetinari's work! But maybe it's time to either move on from libkml or to fork it and make the necessary adaptations into a new upstream. What does Gdal upstream think about it?

**Comment by akrherz** [comment]
FWIW, I found it interesting to see what fedora/epel does with [libkml](https://src.fedoraproject.org/rpms/libkml/blob/rawhide/f/libkml.spec) and the patches it carries to make minizip=1.3 work.

**Comment by h-vetinari** [comment]
I don't know how used libkml is by itself, but it has 1.5mio downloads in conda-forge over the last year, which is quite substantial.

And it's not like our situation is all that great. The current existing landscape, with libkml used in the wild, and now minizip being available as a separate output with the exact same library (and segfaults already observed on feedstocks, as mentioned above), we kinda have to fix this?

I have some good news on that front: though it took some effort, I have a passing test suite (including all the stuff relying on minizip) in #25, with libmkl using only `minizip-static` 1.2 (so no more clashes of `libminizip.so`) 🥳 

It would probably take some deeper review than usual (e.g. I saw that @rouault who commented here yesterday [worked](https://github.com/zlib-ng/minizip-ng/blob/1.2/unzip.h#L10-L11) on minizip a long time ago...?), but it should be possible to fix. I'd also happily send the patches upstream of course (or provide them to a fork, or to gdal or whoever).

**Comment by rouault** [comment]
> But maybe it's time to either move on from libkml or to fork it and make the necessary adaptations into a new upstream. What does Gdal upstream think about it?

The GDAL LIBKML libkml based driver is more advanced than the KML one (that just depends on Expat). That said it wouldn't be the end of the world if the GDAL conda-forge build no longer used libkml.

It probably makes sense to fork libkml/libkml (unless someone reaches to https://www.linkedin.com/in/rashadkm/ to ask him if he can give commit rights). I looked at the Debian libkml package, but they have very few patches (https://salsa.debian.org/debian-gis-team/libkml/-/tree/master/debian/patches) and use their minizip-dev package, but this probably works because they are still on minizip 1.2.13.
I'd still be worried of clashes at runtime between a minizip-static 1.2 (which should probably use zlib's minizip, rather than the one from sourceforge)  and a minizip 4.0, if they share symbols of the same name (it looks minizip-ng has a compatibility layer for zlib's minizip, so I guess there are common names). Based on past experience with mixing libtiff 3.x and 4.0, those issues are very sensitive on the linking order (on Linux).
Another option for libkml would be to use some #define trickery for its internal minizip to rename the symbols, and avoid clashing with the external minizip-ng, the drawback being obviously that it may miss security updates.
TBH I'm a bit lost in all this...

**Comment by h-vetinari** [comment]
> minizip-static 1.2 (which should probably use zlib's minizip, rather than the one from sourceforge)

It's not from source-forge, but from https://github.com/zlib-ng/minizip-ng (the `1.2` branch), which is also where the ancient [page](https://www.winimage.com/zLibDll/minizip.html) of Gilles Vollant refers to nowadays. This is is a superset of what's in zlib's [minizip](https://github.com/madler/zlib/tree/master/contrib/minizip), in particular `ioapi_mem.{c,h}` is missing in zlib's version, which is the API that libkml uses (or rather, libkml uses a very early version of what made it into minizip-ng).

> I'd still be worried of clashes at runtime between a minizip-static 1.2 and a minizip 4.0, if they share symbols of the same name (it looks minizip-ng has a compatibility layer for zlib's minizip, so I guess there are common names).

`minizip-static` 1.2 would not be present at runtime, so there's no risk of symbol collisions. The question is whether libkml exposes anything of the minizip ABI (binary interfact) in its own ABI. That would happen e.g. if the libkml API requires passing a minizip-defined struct to one of libkml's functions. If the same function gets with different structs (e.g. due to another consumer having compiled for a different minizip version), things to boom.

I haven't audited this in detail, but AFAICT, this is not the case.

> TBH I'm a bit lost in all this...

Same, but I wrote down a lot of what I learned while researching this situation [above](https://github.com/conda-forge/libkml-feedstock/issues/2#issuecomment-1746247426), perhaps that helps a bit.

**Comment by ocefpaf** [comment]
> 1.5mio downloads in conda-forge over the last year, which is quite substantial.

That's all gdal pulling it as a dependency. I don't think libkml is used by itself or as a dependency in other projects. So the question is: are gdal users using libkml and how well would we be without it. I think that Even answered that above.

TL;DR if you want to pursue this go for it @h-vetinari ! But if you want to throw the towel I think we are ok dropping libkml as a dependency in gdal.

**Comment by h-vetinari** [comment]
I think the decisions are mostly orthogonal, in the sense that my interest with #25 is to avoid existing packages from experiencing segfaults. Whether gdal drops libkml in the future because it isn't necessary anymore is an unrelated question (except that it becomes a choice rather than being forced by libkml being broken).
